### PR TITLE
Fix wrong assembly used xamlparser

### LINKED ISSUE #14463: [Bug]
### Description
Using CreateContent() on a DataTemplate within the shared Xamarin.Forms project throws a XamlParseException.

## Pull request
I've build Xamarin.Forms and stepped through the source.
The problem lies in the [CreateValueVisitor.cs](https://github.com/xamarin/Xamarin.Forms/blob/809e75547cec1a4c84a2eabdcfddb4b4f43fccd5/Xamarin.Forms.Xaml/CreateValuesVisitor.cs#L46), which uses `Context.RootElement?.GetType().GetTypeInfo().Assembly` instead of `Context.RootAssembly`. 
Using the RootAssembly gives the correct assembly, `MWE` in this case, using the typeinfo.assembly gives `Xamarin.Forms.Core` as the assembly. In the same file at line 172 and 191 this code `var assemblyName = (Context.RootAssembly ?? Context.RootElement?.GetType().GetTypeInfo().Assembly)?.GetName().Name;` uses a check to see if the RootAssembly is set.
I know too little about Xaml to know if using this check is desirable or that this intended behaviour? 
[In the MUAI.NET source I can also find the same "wrong" usage of the RootAssembly.](https://github.com/dotnet/maui/blob/main/src/Controls/src/Xaml/CreateValuesVisitor.cs#L46). 
I am happy to create a pull request for both Xamarin.Forms and MAUI.net. I will need a little guidance on what is needed regarding unittest, since I don't know how I would test this.

### Steps to Reproduce

1. Download this minimal working example: [GitHub](https://github.com/GertjanBrouwer/Xamarin.Forms-Bug-MWE)
2. Put a breakpoint on line 41 of datatemplatecontrol.cs, this is the catch statement of the CreateContent. 
3. Run on an Android Emulator or tablet.

### Expected Behavior
The custom control is shown

### Actual Behavior
An error is thrown.

### Basic Information

- Version with issue: Xamarin.Forms 5.0.0.2012
- Last known good version: Not known
- Platform Target Frameworks: 
  - Android: R 11.0 - API 30

### Environment
<details><summary>Show/Hide Visual Studio info</summary>

```
Microsoft Visual Studio Community 2019
Version 16.9.4
VisualStudio.16.Release/16.9.4+31205.134
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

Visual C++ 2019   00435-60000-00000-AA146
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.9.693.2781
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2112521+5741df381174d72f08e3632bb99f52e8635b6a1a
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.9.693.2781
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.9.693.2781
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.9.693.2781
Azure Functions and Web Jobs Tools

C# Tools   3.9.0-6.21160.10+59eedc33d35754759994155ea2f4e1012a9951e3
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

JetBrains ReSharper Ultimate 2020.1.3   Build 201.0.20200515.122017
JetBrains ReSharper Ultimate package for Microsoft Visual Studio. For more information about ReSharper Ultimate, visit http://www.jetbrains.com/resharper. Copyright © 2021 JetBrains, Inc.

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.40331.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.9.7 (df23ba6)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.9.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

SmartPaster2013   1.0
Paste clipboard into comments or string builder

SQL Server Data Tools   16.0.62103.10080
Microsoft SQL Server Data Tools

Syntax Visualizer   1.0
An extension for visualizing Roslyn SyntaxTrees.

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.30201.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.9.0-6.21160.10+59eedc33d35754759994155ea2f4e1012a9951e3
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.9.0-beta.21102.9+7ce7132f1459095e635194d09d6f73265352029a
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Foo   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VsVim   2.8.0.0
VsVim is a Vim emulator for Visual Studio

Xamarin   16.9.000.273 (d16-9@1bba9e0)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.9.0.318 (remotes/origin/7b35adcdd1b204bbdcb279e0f029d151a20a4bf3@7b35adcdd)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.9.72 (426ebf6)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   11.2.2.1 (d16-9/877f572)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 5e9cb6d
    Java.Interop: xamarin/java.interop/d16-9@54f8c24
    ProGuard: Guardsquare/proguard/v7.0.1@912d149
    SQLite: xamarin/sqlite/3.34.1@daff8f4
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-9@d210f11


Xamarin.iOS and Xamarin.Mac SDK   14.14.2.5 (3836759d4)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

</details>

### Build Logs
```
{Xamarin.Forms.Xaml.XamlParseException: Position 372:26. Type controls:IconButton not found in xmlns clr-namespace:Universal_ONE.Views.Controls
  at Xamarin.Forms.Xaml.CreateValuesVisitor.Visit (Xamarin.Forms.Xaml.ElementNode node, Xamarin.Forms.Xaml.INode parentNode) [0x0005f] in D:\a\1\s\Xamarin.Forms.Xaml\CreateValuesVisitor.cs:55 
  at Xamarin.Forms.Xaml.ElementNode.Accept (Xamarin.Forms.Xaml.IXamlNodeVisitor visitor, Xamarin.Forms.Xaml.INode parentNode) [0x000ac] in D:\a\1\s\Xamarin.Forms.Xaml\XamlNode.cs:158 
  at Xamarin.Forms.Xaml.ElementNode.Accept (Xamarin.Forms.Xaml.IXamlNodeVisitor visitor, Xamarin.Forms.Xaml.INode parentNode) [0x00079] in D:\a\1\s\Xamarin.Forms.Xaml\XamlNode.cs:154 
  at Xamarin.Forms.Xaml.ElementNode.Accept (Xamarin.Forms.Xaml.IXamlNodeVisitor visitor, Xamarin.Forms.Xaml.INode parentNode) [0x00079] in D:\a\1\s\Xamarin.Forms.Xaml\XamlNode.cs:154 
  at Xamarin.Forms.Xaml.ElementNode.Accept (Xamarin.Forms.Xaml.IXamlNodeVisitor visitor, Xamarin.Forms.Xaml.INode parentNode) [0x00079] in D:\a\1\s\Xamarin.Forms.Xaml\XamlNode.cs:154 
  at Xamarin.Forms.Xaml.ElementNode.Accept (Xamarin.Forms.Xaml.IXamlNodeVisitor visitor, Xamarin.Forms.Xaml.INode parentNode) [0x00079] in D:\a\1\s\Xamarin.Forms.Xaml\XamlNode.cs:154 
  at Xamarin.Forms.Xaml.ApplyPropertiesVisitor+<>c__DisplayClass44_0.<SetTemplate>b__0 () [0x000b7] in D:\a\1\s\Xamarin.Forms.Xaml\ApplyPropertiesVisitor.cs:785 
  at Xamarin.Forms.ElementTemplate.CreateContent () [0x00026] in D:\a\1\s\Xamarin.Forms.Core\ElementTemplate.cs:82 
  at Universal_ONE.Resources.Templates.RobotProgramNodeSettingsView.CreateTemplateForItem (System.Object item, Xamarin.Forms.DataTemplateSelector templateSelector) [0x00092] in C:\Users\user\source\repos\Universal ONE\Universal ONE\Resources\Templates\RobotProgramNodeSettingsView.cs:87 
  at Universal_ONE.Resources.Templates.RobotProgramNodeSettingsView.BuildItem () [0x0001b] in C:\Users\user\source\repos\Universal ONE\Universal ONE\Resources\Templates\RobotProgramNodeSettingsView.cs:52 }
```

### Reproduction Link
[Minimal working example](https://github.com/GertjanBrouwer/Xamarin.Forms-Bug-MWE)

### Workaround
I have a simple workaround for this issue. 
1. Find the assembly your custom control is in. I do this by running: `System.Reflection.Assembly.GetExecutingAssembly().GetName().Name;` before the `CreateContent()` call. In this case giving me MWE as an assembly name.
2. Add this assembly to the import, thus giving you this line: `xmlns:controls="clr-namespace:MWE.Controls;assembly=MWE"` in the datatemplate.xaml.

### PR DESCRIPTION
**Pull request by GertjanBrouwer**
### Description of Change ###

The wrong assembly is selected when visiting a custom control node in a DataTemplate inside a ResourceDictionary.

### Issues Resolved ### 

- fixes #14463 

### API Changes ###
 None

### Platforms Affected ### 
- Core/XAML (all platforms)

### Behavioral/Visual Changes ###
None

### Before/After Screenshots ### 
Not applicable

### Testing Procedure ###
I have manually tested this. However, I don't have the resources to run any UI or unittests.

### PR Checklist ###
- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by GertjanBrouwer** [comment]
I can't seem to pull upstream changes from Xamarin.Forms. 
Thus I have deletion in an Azure pipeline in this pull request, can't fix this.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -27,34 +27,6 @@ resources:
       name: xamarin/yaml-templates
       endpoint: xamarin
 
-trigger:
```
Comment (jsuarezruiz) [inline azure-pipelines.yml:30]: Could you revert this unnecessary changes?. If need help reverting etc just let me know.

**Review by jsuarezruiz — COMMENTED** [review]
Could you rebase your branch to fix some conflicts?. Thanks.

**Comment by jsuarezruiz** [comment]
@GertjanBrouwer Could you rebase 5.0.0 branch to fix the conflicts?. Thanks!.

**Comment by GertjanBrouwer** [comment]
> @GertjanBrouwer Could you rebase 5.0.0 branch to fix the conflicts?. Thanks!.

Hi @jsuarezruiz, I did this fix on my work pc at my internship. I currently don't work there anymore so I don't have the environment set up. Do you mind rebasing youself? thank you

**Comment by jfversluis** [comment]
Thanks for the time and effort on this! At this point in the Xamarin.Forms lifecycle I'm not really comfortable making changes like these unless there is a really good indicator that this impacts a lot of people. Seeing that there has been almost no activity on the issue on the PR, I think I'm going to leave this for now.

Still: heel erg bedankt Gertjan!
